src/legacy/core_plugins/kibana/public/discover/np_ready/angular/context.js

```diff
@@ -68,53 +66,81 @@ getAngularModule().config($routeProvider => {
     });
 });
 
-function ContextAppRouteController(
-  $routeParams,
-  $scope,
-  AppState,
-  config,
-  $route,
-  getAppState,
-  globalState
-) {
+function ContextAppRouteController($routeParams, $scope, config, $route) {
   const filterManager = getServices().filterManager;
-  const filterStateManager = new FilterStateManager(globalState, getAppState, filterManager);
   const indexPattern = $route.current.locals.indexPattern.ip;
+  const {
+    startSync: startStateSync,
+    stopSync: stopStateSync,
+    appState,
+    globalState,
+    getAppFilters,
+    getGlobalFilters,
+  } = getState(
+    config.get('context:defaultSize'),
+    indexPattern.timeFieldName,
+    config.get('state:storeInSessionStorage')
+  );
+  this.state = { ...appState.getState() };
+  this.filters = [...getGlobalFilters(), ...getAppFilters()];
+  this.anchorId = $routeParams.id;
+  this.indexPattern = indexPattern;
+  this.discoverUrl = getServices().chrome.navLinks.get('kibana:discover').url;
 
-  this.state = new AppState(createDefaultAppState(config, indexPattern));
-  this.state.save(true);
+  filterManager.setFilters(_.cloneDeep(this.filters));
+  startStateSync();
 
+  // take care of changes in State/URL
+  appState.subscribe(newState => {
+    if (newState && !_.isEqual(this.state, newState)) {
+      this.state = newState;
+      this.filters = [...getGlobalFilters(), ...newState.filters];
+    }
+  });
+  globalState.subscribe(newState => {
+    if (newState && Array.isArray(newState.filters)) {
+      this.filters = [...newState.filters, ...getAppFilters()];
+    } else {
+      this.filters = getAppFilters();
+    }
+  });
+
+  // take care of parameter changes in UI
   $scope.$watchGroup(
     [
       'contextAppRoute.state.columns',
       'contextAppRoute.state.predecessorCount',
       'contextAppRoute.state.successorCount',
     ],
-    () => this.state.save(true)
+    newValues => {
+      const [columns, predecessorCount, successorCount] = newValues;
+      if (Array.isArray(columns) && predecessorCount >= 0 && successorCount >= 0) {
+        appState.set({
+          ...this.state,
+          ...{ columns, predecessorCount, successorCount },
+        });
+      }
+    }
   );
 
-  const updateSubsciption = subscribeWithScope($scope, filterManager.getUpdates$(), {
+  // take care of filtermanager changes
+  const updateSubscription = subscribeWithScope($scope, filterManager.getUpdates$(), {
     next: () => {
-      this.filters = _.cloneDeep(filterManager.getFilters());
+      const appFiltersState = getAppFilters();
+      const appFilters = filterManager.getAppFilters();
+      if (!_.isEqual(appFiltersState, appFilters)) {
```

Comment:
Consider using `compareFilters` instead of regular `isEqual`. The problem is that filter contain formatter function which could have different reference. 
e.g.

```ts
import { compareFilters, COMPARE_ALL_OPTIONS } from '../../../../../../plugins/data/public';

const globalFilterChanged = !compareFilters(
        this.filterManager.getGlobalFilters(),
        globalFilters,
        COMPARE_ALL_OPTIONS
      );
      const appFilterChanged = !compareFilters(
        this.filterManager.getAppFilters(),
        appFilters,
        COMPARE_ALL_OPTIONS
      ); 
```

Reply:
thx, now using `esFilters.compareFilters`